src/jit/passes/constant_propagation_pass.c

```diff
@@ -1,286 +1,90 @@
-      fold_masks[OP_##op] = mask;                       \
-    }                                                   \
-  } op##_init;                                          \
-  template <typename R = struct ir_valueInfo<VALUE_V>,  \
-            typename A0 = struct ir_valueInfo<VALUE_V>, \
-            typename A1 = struct ir_valueInfo<VALUE_V>> \
-  void Handle##op(struct ir *ir, Instr *instr)
-
-// registers a fold callback for the specified signature
-#define REGISTER_FOLD(op, r, a0, a1)                                       \
-  static struct _cpp_##op##_##r##_##a0##_##a1##_init {                     \
-    _cpp_##op##_##r##_##a0##_##a1##_init() {                               \
-      fold_cbs[CALLBACK_IDX(OP_##op, VALUE_##r, VALUE_##a0, VALUE_##a1)] = \
-          &Handle##op<struct ir_valueInfo<VALUE_##r>,                      \
-                      struct ir_valueInfo<VALUE_##a0>,                     \
-                      struct ir_valueInfo<VALUE_##a1>>;                    \
-    }                                                                      \
-  } cpp_##op##_##r##_##a0##_##a1##_init
-
-// common helpers for fold functions
-#define ARG0() (instr->arg[0]->*A0::fn)()
-#define ARG1() (instr->arg[1]->*A1::fn)()
-#define ARG2() (instr->arg[2]->*A1::fn)()
-#define ARG0_UNSIGNED() static_cast<typename A0::unsigned_type>(ARG0())
-#define ARG1_UNSIGNED() static_cast<typename A1::unsigned_type>(ARG1())
-#define ARG2_UNSIGNED() static_cast<typename A1::unsigned_type>(ARG2())
-#define RESULT(expr)                                                           \
-  ir_replace_uses(instr, ir_alloc_constant(                                    \
-                             ir, static_cast<typename R::signed_type>(expr))); \
-  ir_remove_instr(instr)
-
-static FoldFn GetFoldFn(Instr *instr) {
-  auto it = fold_cbs.find(
-      CALLBACK_IDX(instr->op, instr->type,
-                   instr->arg[0] ? (int)instr->arg[0]->type : VALUE_V,
-                   instr->arg[1] ? (int)instr->arg[1]->type : VALUE_V));
-  if (it == fold_cbs.end()) {
-    return nullptr;
-  }
-  return it->second;
-}
-
-static int GetFoldMask(Instr *instr) {
-  return fold_masks[instr->op];
-}
-
-static int GetConstantSig(Instr *instr) {
-  int cnst_sig = 0;
-
-  if (instr->arg[0] && ir_is_constant(instr->arg[0])) {
-    cnst_sig |= ARG0_CNST;
-  }
-
-  if (instr->arg[1] && ir_is_constant(instr->arg[1])) {
-    cnst_sig |= ARG1_CNST;
-  }
-
-  if (instr->arg[2] && ir_is_constant(instr->arg[2])) {
-    cnst_sig |= ARG2_CNST;
-  }
-
-  return cnst_sig;
-}
-
-void ConstantPropagationPass::Run(struct ir *ir) {
-  list_for_each_entry_safe(instr, &ir->instrs, struct ir_instr, it) {
-    int fold_mask = GetFoldMask(instr);
-    int cnst_sig = GetConstantSig(instr);
-    if (!fold_mask || (cnst_sig & fold_mask) != fold_mask) {
+    /* Skip instructions which do not perform any operations */
+    if(instr->op == OP_DEBUG_INFO || instr->op == OP_LABEL)
```

Comment:
Instead of this, I'd make the below binary / unary conditionals first check if the instruction stores a result value i.e. `if (instr->result && has constant binary args) {`